kratos/utilities/sensitivity_builder.cpp

```diff
@@ -0,0 +1,627 @@
+    : mpModelPart(&rModelPart), mpResponseFunction(pResponseFunction)
+{
+    KRATOS_TRY;
+    using sensitivity_builder_cpp::ReplaceDeprecatedNameIfExists;
+    ReplaceDeprecatedNameIfExists(Settings, "nodal_sensitivity_variables",
+                                  "nodal_solution_step_sensitivity_variables");
+    ReplaceDeprecatedNameIfExists(Settings, "element_sensitivity_variables",
+                                  "element_data_value_sensitivity_variables");
+    ReplaceDeprecatedNameIfExists(Settings, "element_data_sensitivity_variables",
+                                  "element_data_value_sensitivity_variables");
+    ReplaceDeprecatedNameIfExists(Settings,
+                                  "condition_data_sensitivity_variables",
+                                  "condition_data_value_sensitivity_variables");
+    ReplaceDeprecatedNameIfExists(Settings, "condition_sensitivity_variables",
+                                  "condition_data_value_sensitivity_variables");
+    Parameters default_settings(R"(
+        {
+            "sensitivity_model_part_name": "PLEASE_SPECIFY_SENSITIVITY_MODEL_PART",
+            "nodal_solution_step_sensitivity_variables": [],
+            "element_data_value_sensitivity_variables" : [],
+            "condition_data_value_sensitivity_variables" : [],
+            "build_mode": "static",
+            "nodal_solution_step_sensitivity_calculation_is_thread_safe" : false
+        })");
+    Settings.ValidateAndAssignDefaults(default_settings);
+
+    auto sensitivity_model_part_name =
+        Settings["sensitivity_model_part_name"].GetString();
+    if (sensitivity_model_part_name != "PLEASE_SPECIFY_SENSITIVITY_MODEL_PART")
+    {
+        mpSensitivityModelPart = mpModelPart->pGetSubModelPart(sensitivity_model_part_name);
+    }
+    else
+    {
+        mpSensitivityModelPart = mpModelPart;
+    }
+    using sensitivity_builder_cpp::ArrayOfStringToVector;
+    mNodalSolutionStepSensitivityVariables = ArrayOfStringToVector(
+        Settings["nodal_solution_step_sensitivity_variables"]);
+    mElementDataValueSensitivityVariables = ArrayOfStringToVector(
+        Settings["element_data_value_sensitivity_variables"]);
+    mConditionDataValueSensitivityVariables = ArrayOfStringToVector(
+        Settings["condition_data_value_sensitivity_variables"]);
+    mBuildMode = Settings["build_mode"].GetString();
+    mNodalSolutionStepSensitivityCalculationIsThreadSafe =
+        Settings["nodal_solution_step_sensitivity_calculation_is_thread_safe"].GetBool();
+    KRATOS_CATCH("");
+}
+
+void SensitivityBuilder::CalculateNodalSolutionStepSensitivities(
+    const std::vector<std::string>& rVariables,
+    ModelPart& rModelPart,
+    AdjointResponseFunction& rResponseFunction,
+    double ScalingFactor)
+{
+    KRATOS_TRY;
+    using sensitivity_builder_cpp::CalculateNodalSolutionStepSensitivities;
+    for (auto& r_variable : rVariables)
```

Comment:
for the other types of variables you split up this function once more (extracted the loop over the vector of strings) is there a reason you did not do it for this one?

Reply:
I extract the loop for both nodal solution step value and non historical data value variables. I added one more level in the non historical data values to avoid templated functions in the header file. I decided to template the function for elements and conditions in the cpp file to reduce the number of code paths to test.